table/tables/state_remote.go

```diff
@@ -0,0 +1,268 @@
+	tid int64
+
+	// Output
+	lockType CachedTableLockType
+	lease    uint64
+	err      error
+}
+
+func (op *loadOP) Exec(data *mockStateRemoteData) {
+	op.lockType, op.lease, op.err = data.Load(op.tid)
+	op.Done()
+}
+
+// lockForReadOP is a kind of rmoteTask
+type lockForReadOP struct {
+	sync.WaitGroup
+	// Input
+	tid int64
+	now uint64
+	ts  uint64
+
+	// Output
+	succ bool
+	err  error
+}
+
+func (op *lockForReadOP) Exec(r *mockStateRemoteData) {
+	op.succ, op.err = r.LockForRead(op.tid, op.now, op.ts)
+	op.Done()
+}
+
+// lockForWriteOP is a kind of remote task
+type lockForWriteOP struct {
+	sync.WaitGroup
+	// Input
+	tid int64
+	now uint64
+	ts  uint64
+
+	// Output
+	err      error
+	oldLease uint64
+}
+
+func (op *lockForWriteOP) Exec(data *mockStateRemoteData) {
+	op.oldLease, op.err = data.LockForWrite(op.tid, op.now, op.ts)
+	op.Done()
+}
+
+type mockStateRemoteData struct {
+	data map[int64]*stateRecord
+}
+
+type stateRecord struct {
+	lease    uint64
+	lockType CachedTableLockType
+}
+
+func newMockStateRemoteData() *mockStateRemoteData {
+	return &mockStateRemoteData{
+		data: make(map[int64]*stateRecord),
+	}
+}
+
+func (r *mockStateRemoteData) Load(tid int64) (CachedTableLockType, uint64, error) {
+	record, ok := r.data[tid]
+	if !ok {
+		return CachedTableLockNone, 0, nil
+	}
+	return record.lockType, record.lease, nil
+}
+
+func (r *mockStateRemoteData) LockForRead(tid int64, now, ts uint64) (bool, error) {
+	record, ok := r.data[tid]
+	if !ok {
+		record = &stateRecord{
+			lease:    ts,
+			lockType: CachedTableLockRead,
+		}
+		r.data[tid] = record
+		return true, nil
+	}
+	switch record.lockType {
+	case CachedTableLockNone:
+		// Add the read lock
+		record.lockType = CachedTableLockRead
+		record.lease = ts
+		return true, nil
+	case CachedTableLockRead:
+		// Renew lease for this case.
+		if record.lease < ts {
+			record.lease = ts
+			return true, nil
+		}
+		// Already read locked.
+		return true, nil
+	case CachedTableLockWrite, CachedTableLockIntend:
+		if now > record.lease {
+			// Outdated...clear orphan lock
+			record.lockType = CachedTableLockRead
+			record.lease = ts
+			return true, nil
+		}
+		return false, nil
+	}
+	return false, errors.New("unknown lock type")
+}
+
+func (r *mockStateRemoteData) LockForWrite(tid int64, now, ts uint64) (uint64, error) {
+	record, ok := r.data[tid]
+	if !ok {
+		record.lockType = CachedTableLockWrite
+		record.lease = ts
+		return 0, nil
+	}
+
+	switch record.lockType {
+	case CachedTableLockNone:
+		record.lockType = CachedTableLockWrite
+		record.lease = ts
+		return 0, nil
+	case CachedTableLockRead:
+		if now > record.lease {
+			// Outdated, clear orphan lock and add write lock directly.
+			record.lockType = CachedTableLockWrite
+			record.lease = ts
+			return 0, nil
+		}
+
+		// Change state to intend, prevent renew lease operation.
+		oldLease := record.lease
+		record.lockType = CachedTableLockIntend
+		record.lease = leaseFromTS(ts)
+		return oldLease, nil
```

Comment:
I did not quit understand this part of code ...

Reply:
This is mentioned in the document. If the current read lock has not expired, if you want to add a write lock, you need to add an intention lock first. This is to prevent the read lock from renewing the contract

Reply:
Suppose currently table is locked by a read lock with lease `ts1`. Then a write request (whose `now < ts1`) comes and change the localType to `CachedTableLockIntend`. After this, another write request (whose `now < ts1`) comes. The finally result is that the second write can obtain the write lock without any wait....

Reply:
@lcwangchao You're right.
 So, for the 'intend' state, there are actually two pieces of infromation required:
- The lease of the old read lock (to wait for old read lock timeout)
- The lease of the intend lock itself (in case panic recover)

Combine them into one is a design mistake here.

Reply:
PTAL. Is my modification correct